core/src/SYCL/Kokkos_SYCL_Parallel_Scan.hpp

```diff
@@ -0,0 +1,299 @@
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKO_SYCL_PARALLEL_SCAN_HPP
+#define KOKKO_SYCL_PARALLEL_SCAN_HPP
+
+#include <Kokkos_Macros.hpp>
+#if defined(KOKKOS_ENABLE_SYCL)
+
+namespace Kokkos {
+namespace Impl {
+
+template <class FunctorType, class... Traits>
+class ParallelScanSYCLBase {
+ public:
+  using Policy = Kokkos::RangePolicy<Traits...>;
+
+ protected:
+  using Member       = typename Policy::member_type;
+  using WorkTag      = typename Policy::work_tag;
+  using WorkRange    = typename Policy::WorkRange;
+  using LaunchBounds = typename Policy::launch_bounds;
+
+  using ValueTraits = Kokkos::Impl::FunctorValueTraits<FunctorType, WorkTag>;
+  using ValueInit   = Kokkos::Impl::FunctorValueInit<FunctorType, WorkTag>;
+  using ValueJoin   = Kokkos::Impl::FunctorValueJoin<FunctorType, WorkTag>;
+  using ValueOps    = Kokkos::Impl::FunctorValueOps<FunctorType, WorkTag>;
+
+ public:
+  using pointer_type   = typename ValueTraits::pointer_type;
+  using value_type     = typename ValueTraits::value_type;
+  using reference_type = typename ValueTraits::reference_type;
+  using functor_type   = FunctorType;
+  using size_type      = Kokkos::Experimental::SYCL::size_type;
+  using index_type     = typename Policy::index_type;
+
+ protected:
+  const FunctorType m_functor;
+  const Policy m_policy;
+  pointer_type m_scratch_space = nullptr;
+
+ private:
+  void scan_internal(cl::sycl::queue& q, pointer_type global_mem,
+                     std::size_t size) const {
+    // FIXME_SYCL optimize
+    constexpr size_t wgroup_size = 32;
+    auto n_wgroups               = (size + wgroup_size - 1) / wgroup_size;
+
+    pointer_type group_results;
+    group_results = static_cast<pointer_type>(sycl::malloc(
+        sizeof(value_type) * n_wgroups, q, sycl::usm::alloc::shared));
```

Comment:
I am not quite convinced that this is better here but I can do that. I think that in the end (not in this PR), we might want to manage the memory allocations here rather in an execution space instance.

Reply:
The reason a `unique_ptr` is better is because no matter how the code leaves the scope (exception, early return, multiple return statements, etc.), `sycl::free` is guaranteed to be called.  This is a win for both correctness and makes it easier to reason about the code.

That being said, if this is moving towards a reusable buffer (as the parallel for kernels and parallel reduce result are), it'll change anyway.

Reply:
I introduced `std::unique_ptr` for this in [`f1f4916` (#3577)](https://github.com/kokkos/kokkos/pull/3577/commits/f1f4916324e8791029f6bcd0d9fe638c81f407ef).